### ISSUE TITLE
bfloat16 @ bfloat16 -> float32?

### ISSUE THREAD

**Issue by njzjz** [jax-ml/ml_dtypes#235]
I see different behaviors in NumPy and JAX.

In NumPy,

```py
>>> import numpy as np
>>> a=np.ones((4,4), dtype=ml_dtypes.bfloat16)
>>> a@a
array([[4., 4., 4., 4.],
       [4., 4., 4., 4.],
       [4., 4., 4., 4.],
       [4., 4., 4., 4.]], dtype=float32)
```
The data type of the output is `float32`.

In JAX,

```py
>>> import jax.numpy as jnp
>>> b=jnp.asarray(a)
>>> b@b
Array([[4, 4, 4, 4],
       [4, 4, 4, 4],
       [4, 4, 4, 4],
       [4, 4, 4, 4]], dtype=bfloat16)
```

The data type of the output is `bfloat16`.

I need clarification about this behavior. I don't see any documentation about it, so I am unsure whether it is a bug or a feature.

**Comment by jakevdp** [comment]
The ml_dtypes `dot` implementation outputs `bfloat16`: https://github.com/jax-ml/ml_dtypes/blob/083b34fcd725923d9c588da1e7d23ea581186583/ml_dtypes/_src/custom_float.h#L527-L542

This is what NumPy calls into for `np.matmul`, so something on NumPy's side appears to be doing this implicit conversion.  That said, `np.dot` outputs `bfloat16`:
```python
In [1]: import numpy as np

In [2]: import ml_dtypes

In [3]: x = np.ones((4, 4), dtype=ml_dtypes.bfloat16)

In [4]: np.dot(x, x)
Out[4]: 
array([[4, 4, 4, 4],
       [4, 4, 4, 4],
       [4, 4, 4, 4],
       [4, 4, 4, 4]], dtype=bfloat16)
```
In JAX, the output dtype will always match the input dtype unless otherwise specified.

**Comment by seberg** [comment]
> This is what NumPy calls into for np.matmul, so something on NumPy's side appears to be doing this implicit conversion

Not quite. `matmul` is it's own ufunc, `np.dot` does indeed call into the `DotFunc` arrfunc slot, though (TBH, I think the newer thought was that we should make `dot` call `matmul` instead).
So this can be addressed by implementing the `matmul` gufunc, although `np.dot` would miss out on any improved handling.

EDIT: N.B. as of NumPy 2.5, this is changed now, implementing `matmul` will also make (sanely shaped) `np.dot()` calls work by falling back to the `matmul` ufunc (if the dotfunc isn't implemented).
